include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1124 @@
+                  H);
+#else
+        AssertThrow(false, ExcNotImplemented());
+#endif
+
+        for (unsigned int i=0; i<n_independent_variables; i++)
+          for (unsigned int j=0; j<i+1; j++)
+            {
+              hessian[i][j] = H[i][j];
+              if (i != j)
+                hessian[j][i] = H[i][j]; // Symmetry
+            }
+
+        // Cleanup :-/
+        for (unsigned int i=0; i<n_independent_variables; i++)
+          delete[] H[i];
+        delete[] H;
+        H = nullptr;
+      }
+
+      // === Vector drivers ===
+
+      static void
+      values (const unsigned int             &active_tape_index,
+              const unsigned int             &n_dependent_variables,
+              const unsigned int             &n_independent_variables,
+              const std::vector<scalar_type> &independent_variables,
+              Vector<scalar_type>            &values)
+      {
+        Assert(independent_variables.size() == n_independent_variables,
+               ExcDimensionMismatch(independent_variables.size(),n_independent_variables));
+        Assert(values.size() == n_dependent_variables,
+               ExcMessage("The length of the dependent function vector must "
+                          " be equal to the number of dependent variables."));
+
+        scalar_type *f = new scalar_type[n_dependent_variables]; // Use smart pointers or std_cxx11::array?
+
+#ifdef DEAL_II_WITH_ADOLC
+        ::function(active_tape_index,
+                   n_dependent_variables,
+                   n_independent_variables,
+                   const_cast<scalar_type *>(independent_variables.data()),
+                   f);
+#else
+        AssertThrow(false, ExcNotImplemented());
+#endif
+
+        for (unsigned int i=0; i<n_dependent_variables; i++)
+          values[i] = f[i];
+
+        // Cleanup :-/
+        delete[] f;
+        f = nullptr;
+      }
+
+      static void
+      jacobian (const unsigned int             &active_tape_index,
+                const unsigned int             &n_dependent_variables,
+                const unsigned int             &n_independent_variables,
+                const std::vector<scalar_type> &independent_variables,
+                FullMatrix<scalar_type>        &jacobian)
+      {
+        Assert(independent_variables.size() == n_independent_variables,
+               ExcDimensionMismatch(independent_variables.size(),n_independent_variables));
+        Assert(jacobian.m() == n_dependent_variables,
+               ExcMessage("The jacobian row length must be equal "
+                          "to the number of dependent variables."));
+        Assert(jacobian.n() == n_independent_variables,
+               ExcMessage("The hessian column length must be equal "
+                          "to the number of independent variables."));
+
+        scalar_type **J = new scalar_type*[n_dependent_variables]; // Use smart pointers or std_cxx11::array?
```

Comment:
and here.

Reply:
The last parameter to `::jacobian` must be a `double**`. I couldn't find a way to safely extract an equivalent data type from `FullMatrix` (i.e. everything I tried broke my tests).